### PR TITLE
BUG: fix a regression where tick direction wasn't conserved by Axis.clear()

### PR DESCRIPTION
**Pull request by neutrinoceros**
## PR Summary

This is a quick fix for  #23806, assuming the change in behaviour I reported there wasn't intentional

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [ ] New features are documented, with examples if plot related.
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Review by tacaswell — APPROVED** [review]

**Review by jklymak — APPROVED** [review]
Will hold off merging if @timhoffm wants to argue this is wrong, unless we are pushing 3.6 final out the door imminently...

**Comment by timhoffm** [comment]
It's ok to go back to 3.5 behavior for this. But we need to consider other parameters as well. For example, 3.5 also keeps `tick_params(... width=10)` under `clear()` whereas 3.6rc does not. I suppose this holds for a lot of the tick_params.

**Comment by neutrinoceros** [comment]
@timhoffm I'm happy to extend the scope of my PR as needed. Should I parametrise my test or extend it so it covers many parameters at once ?

**Comment by neutrinoceros** [comment]
@timhoffm done

File: lib/matplotlib/axis.py
Comment (timhoffm) [inline lib/matplotlib/axis.py:849]: This *looks* better, but I'm a bit of a loss what we actually want.

What's the policy for e.g.
- `zorder`
- the grid parameters like `grid_color`

We should have an explicit statement on why we keep some parameters but not others.

**Review by timhoffm — COMMENTED** [review]
I'm sorry this is such a mess. Let me at least recap the history:

- #20149 is the original issue, that started everything
- #20161 was the bugfix for above (v3.4.2)
- resulting in regressions #20721 and #20219.
- these were hot-fixed in #20826 (v3.4.3)
- #22587 (v3.6.0) claims to be a follow-up cleanup which claims to not change the outcome, but apparently we did not have enough tests.

The target should be preserving the 3.4.3/3.5.x state for. Maybe we should back out #22587 for 3.6 and defer to 3.7. It's only supposed to be a code cleanup after all. Otherwise, I think we need even better tests to ensure we're doing consistent and right things.

File: lib/matplotlib/tests/test_axes.py
Comment (neutrinoceros) [reply lib/matplotlib/tests/test_axes.py:7694]: That's for testing zorder. If the plot is otherwise empty we can't detect when it's broken. I'll add a comment to clarify this

File: lib/matplotlib/axis.py
Comment (neutrinoceros) [reply lib/matplotlib/axis.py:849]: I intentionally left out grid params because I have no opinion there, but I agree it should be discussed.

Comment (neutrinoceros) [reply lib/matplotlib/axis.py:849]: `zorder` seemed to fit in the "visibility" theme, for which a consensus appears to have been reached. Please correct me if I am wrong.

**Comment by neutrinoceros** [comment]
> Otherwise, I think we need even better tests to ensure we're doing consistent and right things.

I'm all for that, but could you clarify what you have in mind ?

File: lib/matplotlib/tests/test_axes.py
Comment (neutrinoceros) [reply lib/matplotlib/tests/test_axes.py:7694]: come to think of it:
- I actually forgot to actually test `zorder`
- I doesn't combine nicely with other parameters I'm testing

I'll add a separate test and simplify this one

Comment (neutrinoceros) [reply lib/matplotlib/tests/test_axes.py:7694]: Actually I don't think I understand what `zorder=-1` is *supposed* to achieve, assuming the behaviour in matplotlib 3.5.3 isn't buggy. I would expect that combining `zorder=-1, direction='in'` with an image would effectively hide ticks behind the image, but it's not the case. In order to keep this PR on the conservative side, I'm not going to change the behaviour here unless required.

Comment (timhoffm) [reply lib/matplotlib/tests/test_axes.py]: > I would expect that combining `zorder=-1, direction='in'` with an image would effectively hide ticks behind the image, but it's not the case

I would share that expectation

>  In order to keep this PR on the conservative side, I'm not going to change the behaviour here

👍

**Comment by timhoffm** [comment]
> > Otherwise, I think we need even better tests to ensure we're doing consistent and right things.
>
> I'm all for that, but could you clarify what you have in mind ?

Essentially covering all parameters and making sure that the behavior is consistent before and after #22587 (or 3.5.x and main); i.e. all parameters that we're cleared before should be cleared now, and all parameters that were kept before should be kept now. An additional annoyance to that is that we accept more paramters in `tick_params()` than documented. The documented ones are translated to tick properties, but we also accept tick properties directly.

**Comment by neutrinoceros** [comment]
I absolutely agree with you, I'm just not sure I'll have the time to dig much further into this. I may, but in the mean time, please feel free to add to this PR if you feel that's appropriate.

**Comment by timhoffm** [comment]
The dev call today we have decided to revert the last change. This will continue the behavior for 3.5 to 3.6. For 3.7 we plan a discussion on what `clear()` actually should do.

@neutrinoceros thanks for this PR, I'll mark it as draft as it needs some changes after the rebase, but the tests will continue to be usful. Until we had the discussion there is not much you can do in this PR.

**Comment by neutrinoceros** [comment]
Do you guys wish to keep this open until the discussion is had ? I'm thinking maybe the test could be merged by itself ? I'm fine either way, I'm just trying to cleanup my personal backlog.

**Comment by tacaswell** [comment]
This is unfortunately still in limbo.

**Comment by neutrinoceros** [comment]
It does not seem necessary anymore, so I'll close this. If it ever changes, anyone is free to continue this work and take off where I left.

**Pull request closed without merging**
